tests/test_client.py

```diff
@@ -182,3 +183,56 @@ def test_should_call_api_level_hooks(no_op_provider_client):
     # Then
     api_hook.before.assert_called_once()
     api_hook.after.assert_called_once()
+
+
+def test_provider_events():
+    provider = NoOpProvider()
+    set_provider(provider)
+
+    other_provider = NoOpProvider()
+    set_provider(other_provider, "my-domain")
+
+    provider_details = ProviderEventDetails(message="message")
+    details = EventDetails.from_provider_event_details(
+        provider.get_metadata().name, provider_details
+    )
+
+    def emit_all_events(provider):
+        provider.emit_provider_ready(provider_details)
+        provider.emit_provider_configuration_changed(provider_details)
+        provider.emit_provider_error(provider_details)
+        provider.emit_provider_stale(provider_details)
+
+    spy = MagicMock()
+
+    client = get_client()
+    client.add_handler(ProviderEvent.PROVIDER_READY, spy.provider_ready)
+    client.add_handler(
+        ProviderEvent.PROVIDER_CONFIGURATION_CHANGED, spy.provider_configuration_changed
+    )
+    client.add_handler(ProviderEvent.PROVIDER_ERROR, spy.provider_error)
+    client.add_handler(ProviderEvent.PROVIDER_STALE, spy.provider_stale)
+
+    emit_all_events(provider)
+    emit_all_events(other_provider)
+
+    spy.provider_ready.assert_called_once_with(details)
+    spy.provider_configuration_changed.assert_called_once_with(details)
+    spy.provider_error.assert_called_once_with(details)
+    spy.provider_stale.assert_called_once_with(details)
+
+
+def test_add_remove_event_handler():
+    provider = NoOpProvider()
+    set_provider(provider)
+
+    spy = MagicMock()
+
+    client = get_client()
+    client.add_handler(ProviderEvent.PROVIDER_READY, spy.provider_ready)
+    client.remove_handler(ProviderEvent.PROVIDER_READY, spy.provider_ready)
+
+    provider_details = ProviderEventDetails(message="message")
+    provider.emit_provider_ready(provider_details)
+
+    spy.provider_ready.assert_not_called()
```

Comment:
I think we need some tests to assert that bound handlers support the late binding of providers. When event handlers are added to `clientX`, and `clientX` is associated with `providerX`, they should run if `providerX` emits events. If `clientX` is then re-associated with `providerY`, then `clientX`'s handlers should run only if `providerY` emits events. Note that this applies to the default provider as well (`providerX` might be the default provider).